File: Array/yanghao/java/Algorithm7_2.java
package jdbc.mvp;

public class Algorithm7_2 {
	
	 public ListNode detectCycle(ListNode head) {
		 if (head == null || head.next == null) {
	            return null;
	        }
	        ListNode slow = head, fast = head;        
	        while (fast != null && fast.next != null) {
	            slow = slow.next;
	            fast = fast.next.next;
	            if (slow == fast) {
	                break;
	            }
	        }
	        if (fast == null || fast.next == null) {
	            return null;
	        }

	        slow = head;
	        while (slow != fast) {
	            slow = slow.next;
	            fast = fast.next;
	        }
	        return slow;  
	 }
}

File: Array/yanghao/kotlin/Algorithm7_2.kt
package staudy
class Algorithm7_2 {

    fun kthToLast(head:ListNode?,k:Int):Int?{
        var a = head
        var p = a
        for (i in 1..k){
            p = p?.next
        }

        while (p != null){
            p = p.next
            a = a?.next
        }
        return a?.`val`
    }
}

class ListNode(var `val`: Int,var next: ListNode? = null)